File: golang/max-dot-product-of-two-subsequences/max-dot-product-of-two-subsequences.go
package main

import (
	"fmt"
	"math"
)

func maxDotProduct(nums1 []int, nums2 []int) int {
	n, m := len(nums1), len(nums2)

	dp := make([][]int, n+1)

	for i := range dp {
		dp[i] = make([]int, m+1)
		for j := range dp[i] {
			dp[i][j] = math.MinInt32
		}
	}

	for i := 1; i <= n; i++ {
		for j := 1; j <= m; j++ {
			current := nums1[i-1] * nums2[j-1]

			include := current
			if dp[i-1][j-1] > 0 {
				include += dp[i-1][j-1]
			}

			skipI := dp[i-1][j]
			skipJ := dp[i][j-1]

			dp[i][j] = max(include, skipI, skipJ)
		}
	}

	return dp[n][m]
}

func max(a, b, c int) int {
	if a >= b && a >= c {
		return a
	}

	if b >= a && b >= c {
		return b
	}

	return c
}

func runTest(testNum int, nums1, nums2 []int, expected int) {
	result := maxDotProduct(nums1, nums2)

	status := "FAIL"
	if result == expected {
		status = "PASS"
	}

	fmt.Printf("Test Case %d: %d (Expected: %d) -> %s\n",
		testNum, result, expected, status)
}

func main() {
	// Example 1
	runTest(1,
		[]int{2, 1, -2, 5},
		[]int{3, 0, -6},
		18,
	)

	// Example 2
	runTest(2,
		[]int{3, -2},
		[]int{2, -6, 7},
		21,
	)

	// Example 3 (Negative result required)
	runTest(3,
		[]int{-1, -1},
		[]int{1, 1},
		-1,
	)

	// Edge Case: Single elements
	runTest(4,
		[]int{-5},
		[]int{5},
		-25,
	)

	// Edge Case: Zeros
	runTest(5,
		[]int{0, 0},
		[]int{1, -5},
		0,
	)
}

File: leetcode/max-dot-product-of-two-subsequences.py
class Solution(object):
    def maxDotProduct(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: int
        """

        n, m = len(nums1), len(nums2)

        dp = [[float('-inf')] * (m + 1) for _ in range(n + 1)]

        for i in range(1, n + 1):
            for j in range(1, m + 1):
                current_product = nums1[i - 1] * nums2[j - 1]

                prev_diagonal = dp[i - 1][j - 1]
                option_include = current_product + max(0, prev_diagonal)

                option_skip_i = dp[i - 1][j]

                option_skip_j = dp[i][j - 1]

                dp[i][j] = max(option_include, option_skip_i, option_skip_j)
        
        return int(dp[n][m])
    
if __name__ == "__main__":
    solver = Solution()
    
    def run_test(case_num, nums1, nums2, expected):
        result = solver.maxDotProduct(nums1, nums2)
        status = "PASS" if result == expected else "FAIL"
        print(f"Test Case {case_num}: {result} (Expected: {expected}) -> {status}")

    # Example 1
    # Subsequences: [2, -2] and [3, -6] -> 2*3 + (-2)*(-6) = 6 + 12 = 18
    run_test(1, [2,1,-2,5], [3,0,-6], 18)
    
    # Example 2
    # Subsequences: [3] and [7] -> 21
    run_test(2, [3,-2], [2,-6,7], 21)
    
    # Example 3 (Negative answer case)
    # Subsequences: [-1] and [1] -> -1 (Must pick non-empty)
    run_test(3, [-1,-1], [1,1], -1)

    # Edge Case: Single elements
    run_test(4, [-5], [5], -25)
    
    # Edge Case: All zeros
    run_test(5, [0, 0], [1, -5], 0)